File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    // List to store the result of postorder traversal
    private List<Integer> result = new ArrayList<>();

    /**
     * Performs postorder traversal of a binary tree
     * @param root The root node of the binary tree
     * @return List of node values in postorder sequence (left, right, root)
     */
    public List<Integer> postorderTraversal(TreeNode root) {
        performPostorderDFS(root);
        return result;
    }

    /**
     * Helper method to recursively traverse the tree in postorder
     * @param node The current node being processed
     */
    private void performPostorderDFS(TreeNode node) {
        // Base case: if node is null, return
        if (node == null) {
            return;
        }
      
        // Traverse left subtree first
        performPostorderDFS(node.left);
      
        // Traverse right subtree second
        performPostorderDFS(node.right);
      
        // Process current node last (postorder)
        result.add(node.val);
    }
}

File: 0145-binary-tree-postorder-traversal/0145-binary-tree-postorder-traversal.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

/**
 * Performs postorder traversal of a binary tree
 * @param root - The root node of the binary tree
 * @returns An array containing node values in postorder sequence (left, right, root)
 */
function postorderTraversal(root: TreeNode | null): number[] {
    // Array to store the traversal result
    const result: number[] = [];
  
    /**
     * Helper function to perform depth-first search in postorder
     * @param node - The current node being processed
     */
    const performDFS = (node: TreeNode | null): void => {
        // Base case: if node is null, return
        if (!node) {
            return;
        }
      
        // Traverse left subtree first
        performDFS(node.left);
      
        // Then traverse right subtree
        performDFS(node.right);
      
        // Finally, process the current node
        result.push(node.val);
    };
  
    // Start the traversal from the root
    performDFS(root);
  
    return result;
}